Feedback overhaul (#386)
cholla-hydro/cholla: docs/sphinx/Physics/DiskModel.md

A central ingredient of any galaxy simulation is the physical model of the gravitational potential. This is a complex topic that is relevant in the context of initial conditions **AND** in the context of evolving the simulation over a timestep. The precise treatment of the potential depends on the context **AND** on Cholla's configuration.

It is instructive to first define all possible components of the total potential. Throughout the rest of this page, we will subsequently describe how treatment of these potentials can vary. The total potential is given by: 

:::{math}

\Phi_{\rm tot} = \Phi_{\rm dm} + \Phi_{\rm stars,old} + \Phi_{\rm stars,young} + \Phi_{\rm gas}
:::

where
-  {math}`\Phi_{\rm dm}` corresponds to the dark matter halo  (always a static potential)
-  {math}`\Phi_{\rm stars,old}` corresponds to older stellar populations in the disk (always a static potential)
- {math}`\Phi_{\rm stars,young}` corresponds to the younger stellar populations (computed with self-gravity from particles)
- {math}`\Phi_{\rm gas}` corresponds to the younger stellar populations (computed with self-gravity from particles)

### Primer on Radiative Cooling

> At the time of writing, simulations must always be initialized with CIE cooling (cooling shuts off below 1e4 K) in order to prevent immediate of the gas disk. The current strategy for using other cooling mechanisms. After running the simulation long enough to start driving turbulence in the gas disk (commonly a few 10s of Myr), you can then restart the simulation with a different cooling recipe.

Reviewer: typo: "immediate **collapse** of the..."

Reviewer: Also the following sentence is not complete.

Author: Fixed

### Primer on Feedback and Star Formation:

> In a simulation without particles, we feedback is implemented with a CGOLs like mechanism.

Reviewer: This option is not actually included in this version of cholla, right? My recollection is that the CGOLS feedback code has only ever existed in branches that have not been merged into this repo, in which case, maybe we should not mention it here?

Author: Yes. I think so. (It wasn't clear to me back when I first wrote these docs)

- use these potentials for initializations and for evolving the simulation over a timestep.

:::{important}
[As we noted above](#configurations), when using the standalone Static-Gravity module, the logic implementing the static potential is duplicated between the initial conditions and the source-term calculations; it's the user's responsibility to ensure logic remains consistent in both parts of the code. 
This is **NOT** a concern when using the Self-Gravity module
:::

Consider the parameterization of the potential of a Miyamoto-Nagai potential (this will be important [again](#analytic-self-gravity-estimate), shortly)

:::{math}
\Phi_{\rm MN}(R,z; M,a,b)\equiv \frac{-G M}{\sqrt{R^2 + (a + \sqrt{z^2 + b^2})^2}}.
:::

At the time of writing, we always define {math}`\Phi_{\rm stars,old}(R,z)=\Phi_{\rm MN}(R,z; M_{\rm stars}, R_{\rm stars}, z_{\rm stars})`, where {math}`R` & {math}`z` are cylindrical coordinates, {math}`R_{\rm stars}` & {math}`z_{\rm stars}` are the scale radius/height, and {math}`M_{\rm stars}` is the mass of the disk.

> 

Reviewer: Would it be possible in this section to give a brief description of how these parameters are set? I.e. are they hardcoded in the disk model, are they runtime parameters...?

Author: Its hardcoded (I added a description)

It is instructive to note that {math}`c_s^2 = \partial P / \partial \rho = K \Gamma \rho^{\Gamma - 1} `

The solution is:

:::{math}
:nowrap:

\begin{eqnarray}
\rho(x) &=& \rho(x_0) \exp\left(-\frac{\Phi(x) - \Phi(x_0)}{c_s^2}\right)&,&\ \Gamma = 1 \\
\rho(x) &=& \rho(x_0) \left[ 1+(\gamma - 1) \frac{\Phi(x) - \Phi(x_0)}{c_s^2(x_0)}\right]^{1/(\gamma-1)}&,&\ {\rm otherwise}
\end{eqnarray}

:::

### Halo Gas

Lets consider the halo gas profile.

As noted in the [Schneider & Robertson 2018](https://ui.adsabs.harvard.edu/abs/2018ApJ...860..135S) , "in setting the halo gas distribution the potential is taken to be spherically symmetric." In other words, the halo gas profile uses a potential called {math}`\Phi_{\rm sph}(r) = \Phi_{\rm stars}(R=0,z=r) + \Phi_{\rm halo}(r)` (it **NEVER** includes contributions from self-gravity). We use an isentropic profile:

:::{math}

\rho_{\rm h}(r) = \rho_{\rm h}(r_0)\left[ 1+(\gamma - 1) \frac{\Phi_{\rm sph}(r) - \Phi_{\rm sph}(r_0)}{c_{s,{\rm h}}^2(r_0)}\right]^{1/(\gamma-1)}.

:::

In the above equation

> - Consequently, we define . At the time of writing, this **never** includes contributions from self-gravity. (this definition of self gravity means that we can **NEVER** achieve perfect )

Reviewer: Couple missing phrases.

:::{math}

\Sigma_{\rm SFR}(R) = a * \Sigma_{{\rm gas},0}^{k_s} * \exp(-r_{\rm cyl} / R_{\rm gas-scale-length}).
:::

Essentially we can use this to incremental rate of star-formation {math}`d{\rm SFR}` in the disk between {math}`r_{\rm cyl}` and {math}`(r_{\rm cyl} + dr_{\rm cyl}`). If we consider some duration of time (in practice set by the `t_out` runtime parameter), you can work an expected number of stars formed between  {math}`r_{\rm cyl}` and {math}`(r_{\rm cyl} + dr_{\rm cyl}`) during that duration. It's straight-forward to convert this function into a PDF.

:::{note}
I can definitely elaborate more -- I found the relevant notes on this topic
:::

At startup we sample this PDF to determine the {math}`r_{\rm cyl}` at which all particles are expected to form. We create star-particles at these radii and distribute the "turn-on times" from a time a little before we start the simulation until the time specified by the `tout` parameter. We use Poisson sampling to distribute these "turn-on times," to target a particular SFR. 

> 

Reviewer: Similar comment here about how the SFR is set. Is this a runtime parameter, or hardcoded somewhere?

Author: Its hardcoded. I added a brief description

You also need to consider 2 simulation phases: initial-conditions and evolution over a timestep.

### Primer on Gravitational Potentials

A central ingredient of any galaxy simulation is the physical model of the gravitational potential. This is a complex topic that is relevant in the context of initial conditions **AND** in the context of evolving the simulation over a timestep. The precise treatment of the potential depends on the context **AND** on Cholla's configuration.

It is instructive to first define all possible components of the total potential. Throughout the rest of this page, we will subsequently describe how treatment of these potentials can vary. The total potential is given by: 

:::{math}

\Phi_{\rm tot} = \Phi_{\rm dm} + \Phi_{\rm stars,old} + \Phi_{\rm stars,young} + \Phi_{\rm gas}
:::

where
-  {math}`\Phi_{\rm dm}` corresponds to the dark matter halo  (always a static potential)
-  {math}`\Phi_{\rm stars,old}` corresponds to older stellar populations in the disk (always a static potential)
- {math}`\Phi_{\rm stars,young}` corresponds to the younger stellar populations (computed with self-gravity from particles)

> - {math}`\Phi_{\rm gas}` corresponds to the younger stellar populations (computed with self-gravity from particles)

Reviewer: ```suggestion
- {math}`\Phi_{\rm gas}` corresponds to the gas disk (computed with self-gravity from particles)
```

### Primer on Feedback and Star Formation:

We provide a basic particle-based feedback approach.
At the moment, all star particles must be initialized at startup (see below).
In the future, the goal is to introduce self-consistent star formation.

## Configurations

Historically, there have been 3 main physics-module configurations for running an Idealized Galaxy Problem:
1. Standalone Static Gravity:
  - In this mode, standalone static gravity is used and particles are **NOT** supported.
  - Effectively, this mode assumes that {math}`\Phi_{\rm stars,young}=0` and {math}`\Phi_{\rm gas}=0`.
  - **Purpose:** A simpler model that is easy to customize
:::{warning}
This configuration has not been tested for quite a while in the dev branch.
:::

:::{caution}
At the time of writing, the logic modelling the static potential is duplicated between the initial conditions and the source-term calculation.
It is the user's responsibility to ensure that consistent models of the potential are used in both parts of the code.
:::

2. Normal Self Gravity

>   - use the self-gravity module to model full gravitational potential while updating timesteps.

Reviewer: ```suggestion
  - use the self-gravity module to model the full gravitational potential while updating timesteps.
```

## Analytic Gravitational Potentials

In this section, we discuss the analytic formulae used for modeling Gravitational Potentials. It may be instructive to look back at our [definitions of various components](#primer-on-gravitational-potentials).
### Static Potentials

First, let's consider static potentials. 

All configurations of the idealized Galaxy Simulations:
- currently assume that you are using a static Gravitational Potential produced by 2 components: (i) component contributed by older stellar populations {math}`\Phi_{\rm stars-old}` and (ii) a component contributed by dark matter in the halo, {math}`\Phi_{\rm halo}`.
- use these potentials for initializations and for evolving the simulation over a timestep.

:::{important}
[As we noted above](#configurations), when using the standalone Static-Gravity module, the logic implementing the static potential is duplicated between the initial conditions and the source-term calculations; it's the user's responsibility to ensure logic remains consistent in both parts of the code. 
This is **NOT** a concern when using the Self-Gravity module
:::

> Consider the parameterization of the potential of a Miyamoto-Nagai potential (this will be important [again](#analytic-self-gravity-estimate), shortly)

Reviewer: ```suggestion
Consider the parameterization of a Miyamoto-Nagai potential (this will be important [again](#analytic-self-gravity-estimate), shortly)
```

:::{math}
\Phi_{\rm MN}(R,z; M,a,b)\equiv \frac{-G M}{\sqrt{R^2 + (a + \sqrt{z^2 + b^2})^2}}.
:::

At the time of writing, we always define {math}`\Phi_{\rm stars,old}(R,z)=\Phi_{\rm MN}(R,z; M_{\rm stars}, R_{\rm stars}, z_{\rm stars})`, where {math}`R` & {math}`z` are cylindrical coordinates, {math}`R_{\rm stars}` & {math}`z_{\rm stars}` are the scale radius/height, and {math}`M_{\rm stars}` is the mass of the disk.
Currently, these values are hardcoded where we initialize the `ClusteredDiskGalaxy` struct that holds the Milky Way properties (in {repository-file}`src/model/particles/disk_galaxy.cu`).

At the time of writing, {math}`\Phi_{\rm dm}` is always assumed to be an NFW profile, or

:::{math}
\Phi_{\rm dm}(r) = \frac{-G M_{\rm vir}}{r[\ln(1+c) - c/(1+c)]}\ \ln \left(1 +\frac{r}{R_{\rm dm}}\right),
:::

> where {math}`r=\sqrt{R^2+z^2}` is radius in spherical coordinates, {math}`M_{\rm vir}` is the dark matter mass, {math}`c` is halo concentration, and {math}`R_{\rm dm}` is the scale length of the halo.

Reviewer: Could you also add where these values are set?

Author: This is now addressed in PR #524

The primary limitation to this approximation for {math}`\hat{\Phi}_{\rm gas,disk}(R,z)` is that we don't account for the fact that the gas disk is truncated. Truncation has 2 impacts:
- less importantly: the shape of truncation can be important for the shape of the potential near the truncation radius
- more importantly: {math}`\hat{\Phi}_{\rm gas,disk}(R,z)` includes the contributions to the gravitational potential from material outside of the simulation domain (if that sounds "wrong" to you, that's becomes most of your physical intuition comes from spherical symmetry).

## Gas Initial Conditions

The ``"Disk_3D"`` and ``"Disk_3D_particles"`` initial conditions will both initialize the gas disk. Both cases initialize an isothermal gas disk within a halo. The latter also subsequently initializes particles.

Cholla offers machinery to setup an idealized galaxy simulation. Specifically, it initializes properties a gas disk inside of gas halo. The conditions broadly match the description provided by  [Schneider & Robertson 2018](https://ui.adsabs.harvard.edu/abs/2018ApJ...860..135S) (there are some extensions when using self-gravity).

> Ideally, we our initial conditions would be in hydrostatic equilibrium (i.e. the properties of a simulation would not change if we hit "go"). For a handful of reasons, we can't actually achieve initial conditions a handful of reasons (related to the gas in the halo and the boundary conditions), but we pick ICs that are stable.

Reviewer: ```suggestion
Ideally, our initial conditions would be in hydrostatic equilibrium (i.e. the properties of a simulation would not change if we hit "go"). For a handful of reasons, we can't actually achieve perfectly hydrostatic initial conditions (related to the gas in the halo and the boundary conditions), but we pick ICs that are stable.
```

## Particle ICs

Particles are only initialized if you use the ``"Disk_3D_particles"`` initial conditions.

We use the Kennicutt–Schmidt law to determine the distribution of particles with respect to {math}`r_{\rm cyl}`

:::{math}

\Sigma_{SFR}(R) = a * \Sigma_{\rm gas}^{k_s},
:::

where `a` is some arbitrary normalization constant and {math}`k_s` is usually 1.4. We can combine this with the formula for the gas surface density, {math}`\Sigma_{\rm gas}(R) = \Sigma_{{\rm gas},0} \exp(-R / R_{\rm d})`, to get a more detailed formula for star-formation surface density:

:::{math}

\Sigma_{\rm SFR}(R) = a * \Sigma_{{\rm gas},0}^{k_s} * \exp(-r_{\rm cyl} / R_{\rm gas-scale-length}).
:::

Essentially we can use this to incremental rate of star-formation {math}`d{\rm SFR}` in the disk between {math}`r_{\rm cyl}` and {math}`(r_{\rm cyl} + dr_{\rm cyl}`). If we consider some duration of time (in practice set by the `t_out` runtime parameter), you can work an expected number of stars formed between  {math}`r_{\rm cyl}` and {math}`(r_{\rm cyl} + dr_{\rm cyl}`) during that duration. It's straight-forward to convert this function into a PDF.

:::{note}

> I can definitely elaborate more -- I found the relevant notes on this topic

Reviewer: Given the questions that arose the other day, I think that might be a good idea.
